docs/design-documents/20231117-better-processors.md

```diff
@@ -0,0 +1,889 @@
+to structured data before stored in field `.Payload.After`.
+
+```yaml
+processors:
+  id:   example
+  type: set.field
+  settings:
+    field: ".Payload.After"
+    value: '{"hoisted":{{ .Payload.After }}}'
+```
+
+### `field.subset.exclude`
+
+Remove a subset of fields from the record, all other fields are left untouched.
+If a field is excluded that contains nested data, the whole tree will be removed.
+It is not allowed to exclude `.Position` or `.Operation`.
+
+#### Configuration
+
+- `fields` (required) - comma separated fields that should be excluded from the record.
+
+#### Example
+
+Exclude all metadata fields, field "foo" in `.Payload.Before` and field "bar" in `.Key`.
+
+```yaml
+processors:
+  id:   example
+  type: field.subset.exclude
+  settings:
+    fields: ".Metadata,.Payload.Before.foo,.Key.bar"
+```
+
+### `field.subset.include`
+
+Include only a subset of fields in the record.
+If a field is included that contains nested data, the whole tree will be included.
+This processor is not applicable to `.Position` and `.Operation`. The remaining
+record fields are included by default (`.Key`, `.Metadata`, `.Payload.Before`,
+`.Payload.After`) unless the configuration contains a field that is nested in one
+of these record fields. For example, if the user configures to include field
+`.Metadata.foo` then only field `foo` will be included in `.Metadata` while all other
+metadata fields will be removed, the remaining record fields won't be changed
+(`.Key`, `.Payload.Before`, `.Payload.After`).
+
+#### Configuration
+
+- `fields` (required) - comma separated fields that should be included in the record.
+
+#### Example
+
+Include only fields `foo` and `bar` in `.Key`, remove all other fields from `.Key`.
+
+```yaml
+processors:
+  id:   example
+  type: field.subset.include
+  settings:
+    fields: ".Key.foo,.Key.bar"
+```
+
+The following configuration is equivalent to the previous one:
+
+```yaml
+processors:
+  id:   example
+  type: field.subset.include
+  settings:
+    fields: ".Key.foo,.Key.bar,.Payload.Before,.Payload.After,.Metadata"
+```
+
+### `field.rename`
+
+Rename a group of fields. It is not allowed to rename top-level fields (`.Operation`,
+`.Position`, `.Key`, `.Metadata`, `.Payload.Before`, `.Payload.After`), only fields
+that are nested.
+
+#### Configuration
+
+- `mapping` (required) - comma separated key:value pairs, where the key is the target
+  field reference and value is the name of the new field (only name, not the full path).
+
+#### Example
+
+```yaml
+processors:
+  id:   example
+  type: field.rename
+  settings:
+    fields: ".Key.foo:bar,.Payload.Before.id:key,.Metadata.topic:subject"
+```
+
+### `field.flatten`
+
+Flatten takes a field that is structured or a map and flattens all nested values into
+top level values. The keys leading to a nested value are concatenated with a separator,
+by default the separator is a dot (`.`).
```

Comment:
how do we handle this if fields already have special characters? e.g `my.key { v: y}` ->`my\.key.v => y` ?

Reply:
Good question, since we're going with the Go template syntax you can always use `{{ index myMap "my key in a string form"}}`.